### PR TITLE
Add MinimumDeploymentFee

### LINKED ISSUE #2001: Contract deployment price coefficient and minimum value
**Summary or problem description**
A lot of effort went into reducing system fees on Neo 3, one of which is contract creation and upgrade fee. There were various discussions around that topic, but the last actual change to that was made by #771 and even that originally intended to have a bit different approach (100/10 values). So at the moment we charge for storage size of contract's code and manifest, which is fair, there should be that proportion, but at the same time I think we have two problems with it:
 * contracts is not just about storage, their effect is much broader and maintaining them on-chain is a bit harder than just storing some bytes
 * it's not hard to deploy thousands of contracts with minimal GAS cost to doing so, one/two bytes contracts that do nothing are considered to be perfectly valid for the chain

At the same time having a limited set of contracts can be beneficial for the network as it allows for some additional optimizations.

**Do you have any solution you want to propose?**
Although contract deployment shouldn't cost a fortune, there should be some reasonable barrier to doing so, therefore I want to propose two changes:
 * apply a coefficient to current pricing (like 3), so that the price would be a little higher than now
 * more importantly, add some minimum policy value for these fees (30-50-100, I don't have any particular number of choice here) to prevent deploying useless contracts at scale

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Network Policy
- Other: Economic model

### PR DESCRIPTION
**Pull request by erikzhang**
Close #2001

### CODE REVIEW

**Comment by erikzhang** [comment]
@roman-khimov Please check it.

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 1]
```diff
@@ -60,7 +80,10 @@ private ContractState Deploy(ApplicationEngine engine, byte[] nefFile, byte[] ma
             if (manifest.Length == 0 || manifest.Length > ContractManifest.MaxLength)
                 throw new ArgumentException($"Invalid Manifest Length: {manifest.Length}");
 
-            engine.AddGas(engine.StoragePrice * (nefFile.Length + manifest.Length));
+            engine.AddGas(Math.Max(
```
Comment (roman-khimov) [inline src/neo/SmartContract/Native/ManagementContract.cs:83]: `Math.Min`?

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 2]
```diff
@@ -44,6 +50,20 @@ internal override void OnPersist(ApplicationEngine engine)
             }
         }
 
+        [ContractMethod(0_01000000, CallFlags.ReadStates)]
+        private long GetMinimumDeploymentFee(StoreView snapshot)
```
Comment (roman-khimov) [inline src/neo/SmartContract/Native/ManagementContract.cs:54]: And maybe it's better to have this value in `Policy` contract.

File: src/neo/SmartContract/Native/ManagementContract.cs (hunk 1 above)
Comment (roman-khimov) [reply src/neo/SmartContract/Native/ManagementContract.cs:83]: Not really, OK.

File: src/neo/SmartContract/Native/ManagementContract.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo/SmartContract/Native/ManagementContract.cs:54]: I'm not sure. This is a parameter related to contract deployment, and maybe it should be placed here.

File: tests/neo.UnitTests/Extensions/NativeContractExtensions.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ namespace Neo.UnitTests.Extensions
 {
     public static class NativeContractExtensions
     {
-        public static ContractState DeployContract(this StoreView snapshot, UInt160 sender, byte[] nefFile, byte[] manifest, long gas = ApplicationEngine.TestModeGas)
+        public static ContractState DeployContract(this StoreView snapshot, UInt160 sender, byte[] nefFile, byte[] manifest, long gas = 200_00000000)
```
Comment (shargon) [inline tests/neo.UnitTests/Extensions/NativeContractExtensions.cs:13]: it's 100?

```suggestion
        public static ContractState DeployContract(this StoreView snapshot, UInt160 sender, byte[] nefFile, byte[] manifest, long gas = 100_00000000)
```

File: src/neo/SmartContract/Native/ManagementContract.cs (hunk 2 above)
Comment (shargon) [reply src/neo/SmartContract/Native/ManagementContract.cs:54]: I think like @roman-khimov , it's related to the environment. But also it's ok too

**Comment by erikzhang** [comment]
Merge?

**Comment by shargon** [comment]
@Tommo-L proposed to reduce the min fee to 1, what do you think? 

https://github.com/neo-project/neo/issues/2001#issuecomment-745105032

**Review by shargon — APPROVED** [review]
we can adjust the value later

**Review by vncoelho — COMMENTED** [review]
It should be on the Police too.

**Review by vncoelho — COMMENTED** [review]
But if it is just 1 GAS then I think it would be ok here.

**Comment by Tommo-L** [comment]
I see we can adjust it later, but it needs the multi-signatures of committee.  And 100GAS is still higher than Ethereum, which is about 0.04 ~ 0.2 ETH(~ 16GAS - 80GAS)  for a simple ERC20 contract.

**Comment by Tommo-L** [comment]
I think this way will be better.

```cs
DeployFee= baseFee + Factor * engine.StoragePrice * (nefFile.Length + manifest.Length)
```
Like:
baseFee, 1 ~ 10GAS
Factor, 1 ~ 5  


Perhaps we can do a test with the Nep17 template.

**Comment by superboyiii** [comment]
We should not adjust `MinimumDeploymentFee` to such a high level. It's not friendly to developers although I know it could reduce malicious attack. Maybe we could adjust it to 10 GAS since the price for OPcode already makes sense now. @erikzhang

**Comment by Tommo-L** [comment]
The size of NEP17 template is about 4KB, nef 2KB, manifest 2KB, which needs about 4GAS in the old way.
In ethereum, a simple ERC20 contract is about 10KB, which needs 0.034ETH ~ 20$ ~ 14GAS.

**Comment by superboyiii** [comment]
I like the old way. If for malicious attack concern, we could increase that to some extent but should not be so high.

**Pull request merged**